### PR TITLE
feat: support ha mode for local registry

### PR DESCRIPTION
**Pull request by kakaZhou719**
### Describe what this PR does / why we need it
Master0 plays an important role in sealer cluster, it hosts many important components such as sealer builtin registry,
application images launched history and others cluster configuration files. if the master0 node is abnormal, a highly
available solution is required to recover on other master nodes.

### Describe how you did it
1. expose Registry struct to clusterfile which is contains configurations about local registry and remote registry.
2. sync all  registry configuration and registry data  to the registry deploy hosts when init the cluster.
3. set `ha mode` as default value for local registry.
4. bugfix: reset host alias about cluster api server to master's their own IP, after join the new master role.
5. optimized some flags of the sealer cluster command. 

### Describe how to verify it
if not set any registry configs, sealer will use non ha mode as default.
if set: below is my test clusterfile,  this will prepare to deploy registry service on three nodes(172.16.26.162,172.16.26.163,172.16.26.164) with customized domain port, basic auth and insure mode options.

```yaml
apiVersion: sealer.cloud/v2
kind: Cluster
metadata:
  creationTimestamp: null
  name: my-cluster
spec:
  hosts:
  - ips:
    - 172.16.26.162
    roles:
    - master
    ssh: {}
  - ips:
    - 172.16.26.163
    roles:
    - master
    ssh: {}
  - ips:
    - 172.16.26.164
    roles:
    - master
    ssh: {}
  - ips:
    - 172.16.26.165
    roles:
    - node
    ssh: {}
  registry: 
    local_registry:
      ha_mode: true
      domain: kakazhou
      port: 5000 
      username: kakazhou
      password: kakazhou123
  image: registry.cn-qingdao.aliyuncs.com/sealer-io/kubernetes:v1.22.15
  ssh:
    passwd: XXX
    pk: /root/.ssh/id_rsa
    port: "22"
    user: root
```

### CODE REVIEW

**Comment by luckyfengyong** [comment]
Please make sure it won't break backward compatibility and upgrade

File: cmd/sealer/cmd/cluster/apply.go [hunk 1]
```diff
@@ -16,30 +16,26 @@ package cluster
 
 import (
 	"fmt"
-	"net"
 	"os"
 	"path/filepath"
 
+	"github.com/sealerio/sealer/cmd/sealer/cmd/utils"
+
```
Comment (VinceCui) [inline cmd/sealer/cmd/cluster/apply.go:23]: remove empty line.

File: pkg/infradriver/ssh_infradriver.go [hunk 2]
```diff
@@ -181,17 +196,47 @@ func (d *SSHInfraDriver) GetHostEnv(host net.IP) map[string]interface{} {
...
+		d.clusterEnv["RegistryPort"] = regConfig.LocalRegistry.Port
 	}
+
 	return d.clusterEnv
 }
 
+func (d *SSHInfraDriver) GetClusterRegistryConfig() v2.Registry {
+	clusterRegistryConfig := d.cluster.Spec.Registry
+	// configure default cluster registry
+	if clusterRegistryConfig.LocalRegistry == nil {
```
Comment (VinceCui) [inline pkg/infradriver/ssh_infradriver.go:214]: i think maybe we can add a flag named HA for LocalRegistry, if true, will set deployHosts as all masters.

Comment (VinceCui) [reply pkg/infradriver/ssh_infradriver.go:214]: Now, user need manually specify deployhosts, which is incomprehensible and troublesome

File: pkg/runtime/kubernetes/join_masters.go [hunk 3]
```diff
@@ -101,5 +101,16 @@ func (k *Runtime) joinMasters(newMasters []net.IP, master0 net.IP, kubeadmConfig
 		logrus.Infof("Succeeded in joining %s as master", m)
 	}
 
+	// set master IP as new apiserver IP address
```
[the remark this answers is not in the document]
Comment (VinceCui) [reply pkg/runtime/kubernetes/join_masters.go:104]: 这个功能你先回滚吧，我在另一个PR里做了。reset apiserver hostalias这个。

File: cmd/sealer/cmd/cluster/apply.go [hunk 4]
```diff
@@ -99,17 +95,17 @@ func NewApplyCmd() *cobra.Command {
 
 			if extension.Type == v12.AppInstaller {
```
Comment (Starnop) [inline cmd/sealer/cmd/cluster/apply.go:96]: How about extract a public method to determine if it is an app image？

File: cmd/sealer/cmd/cluster/run.go [hunk 5]
```diff
@@ -248,3 +244,27 @@ func loadPluginsFromImage(imageMountInfo []imagedistributor.ClusterImageMountInf
...
+func loadImageMode(infraDriver infradriver.InfraDriver, distributor imagedistributor.Distributor) error {
+	regConfig := infraDriver.GetClusterRegistryConfig()
+	// only support load image to local registry at present
+	if regConfig.LocalRegistry == nil {
+		return nil
+	}
+
+	logrus.Infof("start to apply with mode(%s)", common.ApplyModeLoadImage)
+	deployHosts := infraDriver.GetHostIPListByRole(common.MASTER)
+	master0 := deployHosts[0]
```
Comment (Starnop) [inline cmd/sealer/cmd/cluster/run.go:258]: it will panic here if the length of deploy hosts is zero

File: cmd/sealer/cmd/utils/cluster.go [hunk 6]
```diff
@@ -146,11 +150,46 @@ func constructHost(role string, joinIPs []net.IP, scaleFlags *types.Flags, clust
...
+	cluster := &v2.Cluster{}
+	var masterIPList []net.IP
+	var nodeIPList []net.IP
+
+	for _, node := range nodes.Items {
+		addr := getNodeAddress(node)
+		if addr == nil {
+			continue
 		}
+		if _, ok := node.Labels["node-role.kubernetes.io/master"]; ok {
```
Comment (Starnop) [inline cmd/sealer/cmd/utils/cluster.go:168]: use MasterRoleLabel here

File: cmd/sealer/cmd/utils/cluster.go [hunk 7]
```diff
@@ -146,11 +150,46 @@ func constructHost(role string, joinIPs []net.IP, scaleFlags *types.Flags, clust
...
+		nodeIPList = append(nodeIPList, addr)
 	}
-	return
+	cluster.Spec.Hosts = []v2.Host{{IPS: masterIPList, Roles: []string{common.MASTER}}, {IPS: nodeIPList, Roles: []string{common.NODE}}}
+
+	return cluster, nil
+}
+
+func getNodeAddress(node corev1.Node) net.IP {
+	if len(node.Status.Addresses) < 1 {
```
Comment (Starnop) [inline cmd/sealer/cmd/utils/cluster.go:180]: what will happen if len(node.Status.Addresses)==1?

File: cmd/sealer/cmd/cluster/run-app.go [hunk 8]
```diff
@@ -74,18 +73,20 @@ func NewRunAPPCmd() *cobra.Command {
...
 	//runCmd.Flags().StringSliceVar(&appFlags.LaunchArgs, "args", []string{}, "override default LaunchArgs of clusterimage")
-	return runCmd
+	runAppCmd.Flags().StringVarP(&appFlags.ApplyMode, "applyMode", "m", common.ApplyModeApply, "load images to the specified registry in advance")
+
+	return runAppCmd
 }
 
 func installApplication(appImageName string, launchCmds []string, extension v12.ImageExtension,
-	infraDriver infradriver.InfraDriver, imageEngine imageengine.Interface) error {
+	infraDriver infradriver.InfraDriver, imageEngine imageengine.Interface, createMode string) error {
```
Comment (Starnop) [inline cmd/sealer/cmd/cluster/run-app.go:89]: installMode ？

File: cmd/sealer/cmd/cluster/run.go [hunk 9]
```diff
@@ -148,10 +152,9 @@ func NewRunCmd() *cobra.Command {
 	return runCmd
 }
 
-func createNewCluster(clusterImageName string, infraDriver infradriver.InfraDriver, imageEngine imageengine.Interface, cf clusterfile.Interface) error {
+func createNewCluster(clusterImageName string, infraDriver infradriver.InfraDriver, imageEngine imageengine.Interface, cf clusterfile.Interface, createMode string) error {
```
Comment (Starnop) [inline cmd/sealer/cmd/cluster/run.go:155]: It looks like the `clusterImageName` is unused

File: cmd/sealer/cmd/cluster/run.go [hunk 10]
```diff
@@ -248,3 +244,27 @@ func loadPluginsFromImage(imageMountInfo []imagedistributor.ClusterImageMountInf
 
 	return plugins, nil
 }
+
+// loadImageMode just load container image to local registry
+func loadImageMode(infraDriver infradriver.InfraDriver, distributor imagedistributor.Distributor) error {
+	regConfig := infraDriver.GetClusterRegistryConfig()
+	// only support load image to local registry at present
+	if regConfig.LocalRegistry == nil {
```
Comment (Starnop) [inline cmd/sealer/cmd/cluster/run.go:252]: add TODO here

File: cmd/sealer/cmd/utils/cluster.go (hunk 7 above)
Comment (Starnop) [reply cmd/sealer/cmd/utils/cluster.go:180]: add comment here if the logic is right

File: pkg/infradriver/infradriver.go [hunk 11]
```diff
@@ -17,6 +17,8 @@ package infradriver
 import (
 	"net"
 
+	v2 "github.com/sealerio/sealer/types/api/v2"
+
```
Comment (Starnop) [inline pkg/infradriver/infradriver.go:21]: unused blank

File: pkg/infradriver/ssh_infradriver.go [hunk 12]
```diff
@@ -117,7 +118,36 @@ func NewInfraDriver(cluster *v2.Cluster) (InfraDriver, error) {
 		return nil, fmt.Errorf("no hosts specified")
 	}
 
-	if err := checkAllHostsSameFamily(ret.hosts); err != nil {
+	// check registry config is valid,
+	// make sure external registry domain is valid
+	if cluster.Spec.Registry.ExternalRegistry != nil {
```
Comment (Starnop) [inline pkg/infradriver/ssh_infradriver.go:123]: in the future, we should not distinguish the local registry and external registry

Comment (Starnop) [reply pkg/infradriver/ssh_infradriver.go:123]: add TODO here

**Comment by Starnop** [comment]
set non-ha mode as the default value for the local registry if we support config it, and then we will make sure it won't break backward compatibility and upgrade

File: pkg/runtime/kubernetes/kubeadm/kubeadm_config.go [hunk 13]
```diff
@@ -189,7 +189,9 @@ func NewKubeadmConfig(fromClusterFile KubeadmConfig, fromFile string,
 	conf.APIServer.ExtraArgs[EtcdServers] = getEtcdEndpointsWithHTTPSPrefix(masters)
 	conf.IPVS.ExcludeCIDRs = append(conf.KubeProxyConfiguration.IPVS.ExcludeCIDRs, fmt.Sprintf("%s/32", apiServerVIP))
 	conf.KubeletConfiguration.CgroupDriver = cgroupDriver
-
+	// set cluster image repo,kubeadm will pull container image from this registry.
+	conf.ClusterConfiguration.ImageRepository = imageRepo
```
Comment (kakaZhou719) [inline pkg/runtime/kubernetes/kubeadm/kubeadm_config.go:193]: @VinceCui , here will  affects the configuration of the ackd image. pls review.

File: types/api/v2/cluster_types.go [hunk 14]
```diff
@@ -64,6 +66,44 @@ type HostAlias struct {
 	Hostnames []string `json:"hostnames,omitempty"`
 }
 
+type Registry struct {
+	// LocalRegistry is the sealer builtin registry configuration
+	LocalRegistry *LocalRegistry `json:"local_registry,omitempty"`
```
Comment (VinceCui) [inline types/api/v2/cluster_types.go:71]: json tag should be format localRegistry, don't use '_'

File: pkg/infradriver/ssh_infradriver.go [hunk 15]
```diff
@@ -181,17 +211,39 @@ func (d *SSHInfraDriver) GetHostEnv(host net.IP) map[string]interface{} {
...
+		return d.clusterEnv
 	}
-	if _, ok := d.clusterEnv["RegistryPort"]; !ok {
-		d.clusterEnv["RegistryPort"] = common.DefaultRegistryPort
+
+	// expose RegistryDomain and RegistryPort to env in order others needed.
+	regConfig := d.GetClusterRegistryConfig()
+	if !domainExisted {
+		if regConfig.ExternalRegistry != nil {
+			d.clusterEnv["RegistryDomain"] = regConfig.ExternalRegistry.Domain
```
Comment (VinceCui) [inline pkg/infradriver/ssh_infradriver.go:224]: RegistryDomain, RegistryPort should be const variable.

File: pkg/registry/local.go [hunk 16]
```diff
@@ -326,7 +249,7 @@ func (c *localSingletonConfigurator) configureAccessCredential(hosts []net.IP) e
 	return nil
 }
 
-func (c *localSingletonConfigurator) copy2RemoteHosts(src, dest string, hosts []net.IP) error {
+func (c *localConfigurator) copy2RemoteHosts(src, dest string, hosts []net.IP) error {
```
Comment (VinceCui) [inline pkg/registry/local.go:252]: this maybe a common func of infraDriver.

File: pkg/registry/installer.go [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+	distributor imagedistributor.Distributor) Installer {
+	return &localInstaller{
+		currentDeployHosts: currentDeployHost,
+		infraDriver:        infraDriver,
+		LocalRegistry:      regConfig,
+		distributor:        distributor,
+	}
+}
+
+type localInstaller struct {
```
Comment (VinceCui) [inline pkg/registry/installer.go:61]: localInstaller and local registry configurator, difference? why need them both?

File: types/api/v2/cluster_types.go [hunk 18]
```diff
@@ -64,6 +66,44 @@ type HostAlias struct {
...
+	RegistryConfig
+}
+
+type LocalRegistry struct {
+	RegistryConfig
+	// HAMode indicate that whether local registry will be deployed on all master nodes.
+	// if LocalRegistry is not specified, default value is true.
+	HaMode bool `json:"ha_mode,omitempty"`
+	// InsecureMode indicated that whether the local registry is exposed in HTTPS.
+	// if true sealer will generate default ssl cert.
```
Comment (VinceCui) [inline types/api/v2/cluster_types.go:93]: if true, sealer will NOT generate default ssl cert.

File: pkg/registry/installer.go (hunk 17 above)
Comment (kakaZhou719) [reply pkg/registry/installer.go:61]: Installer provide registry lifecycle management local registry will implement it. Configurator provide registry configuration management, local registry and external registry will implement it.

**Comment by VinceCui** [comment]
LGTM

**Comment by Starnop** [comment]
LGTM

**Review by Starnop — APPROVED** [review]

**Pull request merged**